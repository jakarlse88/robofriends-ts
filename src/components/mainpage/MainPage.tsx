import './MainPage.css';
import CardList from '../cardlist/CardList';
import ErrorBoundary from '../errorboundary/ErrorBoundary';
import Header from '../header/Header';
import * as React from 'react';
import Scroll from '../scroll/Scroll';
import SearchBox from '../searchbox/SearchBox';

interface Robots {
    id: number,
    name: string,
    username: string,
    email: string
};

interface MainPageProps {
    query: string,
    robots: Array<Robots>,
    actions?: any,
    children?: any
};

export default class MainPage extends React.Component<MainPageProps> {
    constructor(props: any) {
        super(props);
    }
    
    filterRobots = (robots: Array<Robots>) => {
        const { query } = this.props;
        const filteredRobots: Array<Robots> = robots.length ?
            robots.filter(robot => {
                return robot.name.toLowerCase().includes(query.toLowerCase());
            })
            : [];
        return filteredRobots;
    }

    render() {
        const {
            actions,
            robots,
        } = this.props;

        return !robots.length ?
            <h1>Loading</h1>
            : (
                <div className='tc'>
                    <Header />
                    <SearchBox
                        searchChange={(e: any) => actions.setQuery(e.target.value)} />
                    <Scroll>
                        <ErrorBoundary>
                            <CardList robots={this.filterRobots(robots)} />
                        </ErrorBoundary>
                    </Scroll>
                </div>
            );
    }
}