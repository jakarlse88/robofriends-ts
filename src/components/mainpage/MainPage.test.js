import { shallow } from 'enzyme';
import MainPage from './MainPage';
import React from 'react';

let wrapper;
beforeEach(() => {
    const mockProps = {
        robots: [],
        query: ''
    };
    wrapper = shallow(<MainPage {...mockProps}/>);
})

describe('MainPage component', () => {
    it('successfully renders the MainPage component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('filters robots correctly', () => {
        const mockProps = {
            robots: [
                {
                    id: 3,
                    name: 'John',
                    email: 'john@gmail.com'
                }
            ],
            query: 'a'
        };
        const wrapper = shallow(<MainPage {...mockProps} />);
        expect(wrapper.instance().filterRobots([])).toEqual([]);
        expect(wrapper.instance().filterRobots(mockProps.robots)).toEqual([]);
    });
})