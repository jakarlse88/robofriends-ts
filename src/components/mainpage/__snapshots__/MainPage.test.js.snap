// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage component successfully renders the MainPage component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainPage
    query=""
    robots={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Loading",
    },
    "ref": null,
    "rendered": "Loading",
    "type": "h1",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Loading",
      },
      "ref": null,
      "rendered": "Loading",
      "type": "h1",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
